wgengine/tsdns/forwarder.go

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsdns
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"hash/crc32"
+	"math/rand"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"inet.af/netaddr"
+	"tailscale.com/types/logger"
+)
+
+// headerBytes is the number of bytes in a DNS message header.
+const headerBytes = 12
+
+// forwardQueueSize is the maximal number of requests that can be pending delegation.
+// Note that this is distinct from the number of requests that are pending a response,
+// which is not limited (except by txid collisions).
+const forwardQueueSize = 64
+
+// connCount is the number of UDP connections to use for forwarding.
+const connCount = 32
+
+const (
+	// cleanupInterval is the interval between purged of timed-out entries from txMap.
+	cleanupInterval = 30 * time.Second
+	// responseTimeout is the maximal amount of time to wait for a DNS response.
+	responseTimeout = 5 * time.Second
+)
+
+var errNoUpstreams = errors.New("upstream nameservers not set")
+
+var aLongTimeAgo = time.Unix(0, 1)
+
+type forwardedPacket struct {
+	payload []byte
+	dst     net.Addr
+}
+
+type forwardingRecord struct {
+	src       netaddr.IPPort
+	createdAt time.Time
+}
+
+// txid identifies a DNS transaction.
+//
+// As the standard DNS Request ID is only 16 bits, we extend it:
+// the lower 32 bits are the zero-extended bits of the DNS Request ID;
+// the upper 32 bits are the CRC32 checksum of the first question in the request.
+// This makes probability of txid collision negligible.
+type txid uint64
+
+// getTxID computes the txid of the given DNS packet.
+func getTxID(packet []byte) txid {
+	if len(packet) < headerBytes {
+		return 0
+	}
+
+	dnsid := binary.BigEndian.Uint16(packet[0:2])
+	qcount := binary.BigEndian.Uint16(packet[4:6])
+	if qcount == 0 {
+		return txid(dnsid)
+	}
+
+	offset := headerBytes
+	for i := uint16(0); i < qcount; i++ {
+		// Note: this relies on the fact that names are not compressed in questions,
```

Comment:
Is this guaranteed by the DNS protocol?

Reply:
My reading of the relevant RFC ([RFC 1035](https://tools.ietf.org/html/rfc1035)) is that it doesn't prohibit it, but it doesn't happen in practice, probably because an implementation doing it would have to be "excessively smart": having multiple questions is a DNS packet appears to be [ill-defined](https://stackoverflow.com/a/4083071), and a single question would have to contain a pointer to an _answer_ to be compressed, which would be "excessively smart" and pointless (the answer can just as well point to the question).

A draft RFC explicitly says, in _summarizing_ RFC 1035:
> It is important that these pointers always point backwards.

That phrase isn't found in RFC 1035, so I'm not sure where it originally comes from.

As for where I got this "fact" from, that was djb's [notes on DNS](https://cr.yp.to/djbdns/notes.html):

> The precise rule is that a name can be compressed if it is a response owner name, the name in NS data, the name in CNAME data, the name in PTR data, the name in MX data, or one of the names in SOA data.

That also doesn't appear to come from the RFC, but it looked like a very confident assertion.

Reply:
As for where I got this "fact" from, that was djb's [notes on DNS](https://cr.yp.to/djbdns/notes.html):

> The precise rule is that a name can be compressed if it is a response owner name, the name in NS data, the name in CNAME data, the name in PTR data, the name in MX data, or one of the names in SOA data.

That also doesn't appear to come from the RFC, but it looked like a very confident assertion.

Reply:
Yeah, that makes sense. I would expand this comment to link to these references. I think the biggest risk would be the possibility that a question record would contain the same label twice (e.g. "foo.foo.bar.com"), and a clever implementation would try to compress the second `foo`. But it seems unlikely enough that this is fine.

Reply:
I believe even that case would be fine since compression can only be applied to proper suffixes:

> an entire domain name or a list of labels at the end of a domain name is replaced with a pointer to a prior occurance of the same name.

(from RFC 1035).

Reply:
So actually only a suffix can get sent to a suffix, since the pointed-to name will be considered in its entirety (until NUL).